File: background-jobs/worker/python/worker.py
import json
import os
import threading
import time
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

interval = int(os.getenv("WORKER_INTERVAL_SECONDS", "30"))
status_port = int(os.getenv("STATUS_PORT", "3000"))

last_run = time.monotonic()

# Reports whether the worker loop is still ticking, not just whether the
# process is running (Bzync Cloud's compute already checks that
# separately via container state + crash-loop detection) — same
# reachable/unreachable JSON convention as the redis/garage/seaweedfs
# templates' status sidecar.
stale_after = interval * 2 + 30


class StatusHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        stale = time.monotonic() - last_run > stale_after
        body = json.dumps(
            {
                "status": "error" if stale else "ok",
                "service": "worker",
                "worker": "unreachable" if stale else "reachable",
            }
        ).encode()
        self.send_response(503 if stale else 200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *args):
        pass  # Don't spam job logs with health check request lines.


def serve_status() -> None:
    ThreadingHTTPServer(("0.0.0.0", status_port), StatusHandler).serve_forever()


threading.Thread(target=serve_status, daemon=True).start()

# This loop is a placeholder heartbeat, safe to scale to any replica count
# as-is since it does nothing but log. If you replace it with real work
# (polling a queue or a database table), make sure each item is claimed by
# exactly one replica before processing it (e.g. `SELECT ... FOR UPDATE SKIP
# LOCKED`, or your broker's own ack/visibility-timeout semantics) — without
# that, N replicas will each pick up and process the same item.
while True:
    print(f"Python worker heartbeat {datetime.now(timezone.utc).isoformat()}", flush=True)
    last_run = time.monotonic()
    time.sleep(interval)

File: background-jobs/worker/go/worker.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strconv"
	"sync/atomic"
	"time"
)

var lastRunUnixNano atomic.Int64

func main() {
	interval := 30 * time.Second
	if s := os.Getenv("WORKER_INTERVAL_MS"); s != "" {
		if n, err := strconv.Atoi(s); err == nil {
			interval = time.Duration(n) * time.Millisecond
		}
	}

	run()
	go serveStatus(statusPort(), interval)

	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for range ticker.C {
		run()
	}
}

// run is a placeholder heartbeat, safe to scale to any replica count as-is
// since it does nothing but log. If you replace it with real work (polling
// a queue or a database table), make sure each item is claimed by exactly
// one replica before processing it (e.g. `SELECT ... FOR UPDATE SKIP
// LOCKED`, or your broker's own ack/visibility-timeout semantics) —
// without that, N replicas will each pick up and process the same item.
func run() {
	fmt.Printf("Go worker heartbeat %s\n", time.Now().UTC().Format(time.RFC3339))
	lastRunUnixNano.Store(time.Now().UnixNano())
}

func statusPort() string {
	if p := os.Getenv("STATUS_PORT"); p != "" {
		return p
	}
	return "3000"
}

// serveStatus reports whether the worker loop is still ticking, not just
// whether the process is running (Bzync Cloud's compute already checks
// that separately via container state + crash-loop detection) — same
// reachable/unreachable JSON convention as the redis/garage/seaweedfs
// templates' status sidecar.
func serveStatus(port string, interval time.Duration) {
	staleAfter := interval*2 + 30*time.Second
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		last := time.Unix(0, lastRunUnixNano.Load())
		w.Header().Set("Content-Type", "application/json")
		if time.Since(last) > staleAfter {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "error", "service": "worker", "worker": "unreachable",
				"lastRun": last.UTC().Format(time.RFC3339),
			})
			return
		}
		_ = json.NewEncoder(w).Encode(map[string]string{
			"status": "ok", "service": "worker", "worker": "reachable",
			"lastRun": last.UTC().Format(time.RFC3339),
		})
	})
	if err := http.ListenAndServe(":"+port, nil); err != nil {
		fmt.Fprintf(os.Stderr, "status server: %v\n", err)
	}
}
